toolchain/check/check.cpp

```diff
@@ -231,14 +250,360 @@ static auto InitPackageScopeAndImports(Context& context, UnitInfo& unit_info)
       << "Created an unexpected number of IRs";
 }
 
+namespace {
+// State used to track the next inline method body that we will encounter and
+// need to reorder.
+struct NextInlineMethodCache {
+  const Parse::Tree* tree;
+  Parse::InlineMethodIndex index = Parse::InlineMethodIndex::Invalid;
+  Parse::NodeId start_id = Parse::NodeId::Invalid;
+
+  // Set the specified inline method index as being the next one to parse.
+  auto SkipTo(Parse::InlineMethodIndex next_index) -> void {
+    index = next_index;
+    if (static_cast<std::size_t>(index.index) ==
+        tree->inline_methods().size()) {
+      start_id = Parse::NodeId::Invalid;
+    } else {
+      start_id = tree->inline_methods().Get(index).start_id;
+    }
+  }
+};
+}  // namespace
+
+// Determines whether we are currently declaring a name in a scope in which
+// method definitions are delayed. When entering another delayed method scope,
+// the inner declaration's methods are parsed at the end of the outer
+// declaration, not the inner one.  For example:
+//
+// ```
+// class A {
+//   class B {
+//     fn F() -> A { return {}; }
+//   }
+// } // A.B.F is parsed here, with A complete.
+//
+// fn F() {
+//   class C {
+//     fn G() {}
+//   } // C.G is parsed here.
+// }
+// ```
+static auto IsInDelayedMethodScope(Context& context) -> bool {
+  auto inst_id = context.name_scopes().GetInstIdIfValid(
+      context.decl_name_stack().PeekTargetScope());
+  if (!inst_id.is_valid()) {
+    return false;
+  }
+  switch (context.insts().Get(inst_id).kind()) {
+    case SemIR::ClassDecl::Kind:
+    case SemIR::ImplDecl::Kind:
+    case SemIR::InterfaceDecl::Kind:
+      // TODO: Named constraints, mixins.
+      return true;
+
+    default:
+      return false;
+  }
+}
+
+// Determines whether this node kind is the start of a delayed method scope.
+static auto IsStartOfDelayedMethodScope(Parse::NodeKind kind) -> bool {
+  switch (kind) {
+    case Parse::NodeKind::ClassDefinitionStart:
```

Comment:
Hm, I hadn't considered that. I don't think we have a way to ask "does this node match that `NodeIdOneOf` type?" that doesn't CHECK-fail if not, currently, but we could add that if you think it would aid the clarity here.

(FWIW, I've been trying to arrange this so that LLVM will smartly combine the three `switch`es per `kind` into a single `switch`. I'm not sure whether that's as likely to happen if we use something other than `switch` here, but we can dig into that if we see performance concerns.)

I want to keep the list of start and end node kinds here, because it's important that these three lists all match, so I wouldn't want to use a typedef declared out in `node_ids.h`, in case that affects your opinion here. This is also why I'm reluctant to use a `NodeCategory` for this instead of `NodeIdOneOf` or an explicit list of kinds.

Reply:
Eh, maybe it's not common enough? I was just trying to think of a way to write shorter cde.